web/src/search_pill.ts

```diff
@@ -89,45 +99,86 @@ function append_user_pill(
+        case "channel":
+            return stream_data.get_sub(term.operand) !== undefined;
+        case "sender":
+        case "dm":
+        case "dm-including":
+            return term.operand
+                .split(",")
+                .every((email) => people.get_by_email(email) !== undefined);
+        default:
+            return true;
+    }
+}
+
 export function set_search_bar_contents(
     search_terms: NarrowTerm[],
     pill_widget: SearchPillWidget,
-    set_search_bar_text?: (text: string) => void,
+    set_search_bar_text: (text: string) => void,
 ): void {
     pill_widget.clear();
     let partial_pill = "";
+    const invalid_inputs = [];
+
     for (const term of search_terms) {
-        if (user_pill_operators.has(term.operator) && term.operand !== "") {
-            const user_emails = term.operand.split(",");
-            const users = user_emails.map((email) => {
-                const user = people.get_by_email(email);
-                assert(user !== undefined);
-                return user;
-            });
-            append_user_pill(users, pill_widget, term.operator, term.negated ?? false);
-            continue;
-        }
         const input = Filter.unparse([term]);
+
         // If the last term looks something like `dm:`, we
         // don't want to make it a pill, since it isn't isn't
         // a complete search term yet.
         // Instead, we keep the partial pill to the end of the
         // search box as text input, which will update the
         // typeahead to show operand suggestions.
-        if (
-            set_search_bar_text !== undefined &&
-            input.at(-1) === ":" &&
-            term.operand === "" &&
-            term === search_terms.at(-1)
-        ) {
+        if (input.at(-1) === ":" && term.operand === "" && term === search_terms.at(-1)) {
             partial_pill = input;
             continue;
         }
+
+        if (!is_valid_pill(term)) {
+            invalid_inputs.push(input);
+            continue;
+        }
+
+        if (user_pill_operators.has(term.operator) && term.operand !== "") {
+            const users = term.operand.split(",").map((email) => {
+                const user = people.get_by_email(email);
+                assert(user !== undefined);
+                return user;
+            });
+            append_user_pill(users, pill_widget, term.operator, term.negated ?? false);
+            continue;
+        }
+
         pill_widget.appendValue(input);
     }
     pill_widget.clear_text();
-    if (set_search_bar_text !== undefined) {
-        set_search_bar_text(partial_pill);
+
+    const search_bar_text_strings = [...invalid_inputs];
+    if (partial_pill !== "") {
+        search_bar_text_strings.push(partial_pill);
+    }
+    set_search_bar_text(search_bar_text_strings.join("  "));
+    if (invalid_inputs.length) {
+        $("#search_query").addClass("shake");
```

Comment:
Why is there a double-space used in this join?

Reply:
That's a mistake! Thanks